R/utilities.R

```diff
@@ -2511,6 +2511,356 @@ quiet <- function(expr, suppressWarnings = FALSE) {
+    stop('\n\U1F6D1 {field_types} cannot be empty.')
+  }
+
+  # Make field_types names lowercase to match data
+  names(field_types) <- tolower(names(field_types))
+
+  if (!identical(sort(tolower(names(ph.data))), sort(tolower(names(field_types))))) {
+    stop('\n\U1F6D1 Conversion requires exactly one TSQL datatype per column name in {ph.data}.')
+  }
+
+  if(!is.logical(validate_before)){
+    stop('\n\U1F6D1 {validate_before} must be specified as a logical (i.e., TRUE, T, FALSE, or F)')
+  }
+
+  if(!is.logical(validate_after)){
+    stop('\n\U1F6D1 {validate_after} must be specified as a logical (i.e., TRUE, T, FALSE, or F)')
+  }
+
+  if(!is.logical(verbose)){
+    stop('\n\U1F6D1 {verbose} must be specified as a logical (i.e., TRUE, T, FALSE, or F)')
+  }
+
+  if(!is.logical(return_log)){
+    stop('\n\U1F6D1 {return_log} must be specified as a logical (i.e., TRUE, T, FALSE, or F)')
+  }
+
+  # IF `validate_before = TRUE` ----
+  if (validate_before) {
+    if (verbose) {
+      message('\U0001f50d Validating field types before conversion...')
+    }
+
+    tryCatch({
+      tsql_validate_field_types(ph.data, field_types) # if there is no error, no need to convert anything
+      if (verbose) {
+        message('\U0001f642 Success! Your desired TSQL data types are already compatible with your dataset. No conversion needed.')
+      }
+
+      if (return_log) {
+        return(list(
+          data = ph.data,
+          conversion_log = data.table(
+            column = names(field_types),
+            original_type = sapply(ph.data, function(x) class(x)[1]),
+            target_r_type = "No conversion needed",
+            target_tsql_type = field_types,
+            conversion_success = TRUE,
+            notes = "Already compatible"
+          )
+        ))
+      } else {
+        return(ph.data)
+      }
+    }, error = function(e) {
+      if (verbose) {
+        message('\u26A0\ufe0f Pre-conversion validation found issues. Continuing with conversion attempt(s)...')
+      }
+    })
+  }
+
+  # Define TSQL to R type mapping ----
+  tsql_to_r_mapping <- list(
+    tinyint = "integer",
+    smallint = "integer",
+    int = "integer",
+    bigint = "numeric", # R integers can't handle all BIGINT
+    decimal = "numeric",
+    numeric = "numeric",
+    float = "numeric",
+    real = "numeric",
+    money = "numeric",
+    smallmoney = "numeric",
+    char = "character",
+    varchar = "character",
+    text = "character",
+    nchar = "character",
+    nvarchar = "character",
+    ntext = "character",
+    bit = "integer",
+    date = "Date",
+    datetime = "POSIXct",
+    datetime2 = "POSIXct",
+    smalldatetime = "POSIXct",
+    datetimeoffset = "POSIXct",
+    binary = "character",
+    varbinary = "character"
+  )
+
+  # Helper function: map_tsql_to_r_class() ----
+  map_tsql_to_r_class <- function(tsql_type) {
+    base_type <- tolower(gsub("\\(.*\\)", "", trimws(tsql_type))) # drop specifiers in (), e.g., CHAR(12) > char
```

Comment:
Do you actually want to do that? What if you're trying to upload a 13 character string to a char(12)?

Reply:
This is for the first pass, just to make sure it is the correct type, to see if it is convertible.
Assuming people leave the default `validate_after = TRUE`, then it will warn that some values for specific columns are outside of the acceptable limit. So, first, is is the right type, then does it meet the parameters for length/size.